src/algebra/non_associative/basic.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.algebra.basic
+
+/-!
+# Non-associative, non-unital algebras
+
+This files defines non-associative, non-unital algebras together with their morphisms and basic
+lemmas to make them usable.
+
+## Main definitions
+
+  * `na_semiring`
+  * `na_algebra`
+  * `na_algebra_hom`
+
+## Tags
+
+non-associative, non-unital, algebra
+-/
+
+universes u₁ u₂ u₃ u₄
+
+variables (R : Type u₁) (A : Type u₂) (B : Type u₃) (C : Type u₄) [semiring R]
+
+set_option old_structure_cmd true
+
+/-- A not-necessarily-unital, not-necessarily-associative semiring, or "non-associative ring" for
+short. -/
+class na_semiring extends add_comm_monoid A, distrib A, mul_zero_class A
+
+-- https://github.com/leanprover-community/lean/issues/285
+set_option old_structure_cmd false
+
+/-- A not-necessarily-unital, not-necessarily-associative algebra, or "non-associative algebra" for
+short. -/
+@[protect_proj]
+class na_algebra [na_semiring A] extends semimodule R A :=
```

Comment:
If `A` is a comm_semiring`, are these axioms sufficient to reconstruct an `algebra`? That is to say, what assumptions do you need to define `na_algebra.to_algebra`?

Reply:
To define `na_algebra.to_algebra` we need:

- Commutative coefficients `R`
- A unit in `A`
- Associative multiplication in `A`

Reply:
`[comm_semiring A]` provides the last two, and I think the axioms are at least enough to show that `(s * t) • a = (t * s) • a`, but I guess that's not quite as strong as showing that `s * t = t * s`

Reply:
I think what you are thinking of might be:
```lean
def na_algebra.to_algebra
  (R A : Type*) [comm_semiring R] [semiring A] [na_algebra R A] : algebra R A :=
sorry
```

Reply:
That's certainly one of the things I was thinking, yes. At that point; there's only one difference between `na_algebra` and `algebra`; the ability to control the definitional reduction of `algebra_map`.

Reply:
I've added a commit along these lines.